Build failure: qimgv

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

`qimgv` fails when `cudaSupport = true`.

```sh
nix build -f ./. --arg config '{ cudaSupport = true; allowUnfree = true;  }' -L qimgv
```

### Can Hydra reproduce this build failure?

No, Hydra cannot reproduce this build failure.

### Link to Hydra build job

_No response_

### Relevant log output

```console
Running phase: qtPreHook
@nix { "action": "setPhase", "phase": "qtPreHook" }
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/bsipnp1jam0wpbsgpk1b5zjsxi5y4b0y-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
fixing cmake files...
cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/share/doc/qimgv -DCMAKE_INSTALL_INFODIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/include -DCMAKE_INSTALL_SBINDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/4ijy8jbsiqmj37avrk83gn2m903486mr-gcc-wrapper-14-20241116/bin/strip -DCMAKE_RANLIB=/nix/store/4ijy8jbsiqmj37avrk83gn2m903486mr-gcc-wrapper-14-20241116/bin/ranlib -DCMAKE_AR=/nix/store/4ijy8jbsiqmj37avrk83gn2m903486mr-gcc-wrapper-14-20241116/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/jq8f871phxy97y6rlzm2jn24mipi60ap-qimgv-1.0.3-unstable-2024-10-11 -DVIDEO_SUPPORT=ON
-- The CXX compiler identification is GNU 14.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/4ijy8jbsiqmj37avrk83gn2m903486mr-gcc-wrapper-14-20241116/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build configuration: Release
-- Qt Version: 5.15.16
-- Found PkgConfig: /nix/store/zbpskag55c4qvpzjvwrlxqs40lj65mwn-pkg-config-wrapper-0.29.2/bin/pkg-config (found version "0.29.2")
-- Checking for module 'exiv2'
--   Found exiv2, version 0.28.4
CMake Warning (dev) at /nix/store/xpx3wyc4ws1blif4y1zz92ys8gf2y3wz-opencv-4.9.0/lib/cmake/opencv4/OpenCVConfig.cmake:86 (find_package):
  Policy CMP0146 is not set: The FindCUDA module is removed.  Run "cmake
  --help-policy CMP0146" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

Call Stack (most recent call first):
  /nix/store/xpx3wyc4ws1blif4y1zz92ys8gf2y3wz-opencv-4.9.0/lib/cmake/opencv4/OpenCVConfig.cmake:108 (find_host_package)
  CMakeLists.txt:62 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /nix/store/wjb1rq9xgxzq6n1b7pbrbsk5lrdn9qc7-cmake-3.31.3/share/cmake-3.31/Modules/FindCUDA.cmake:883 (message):
  Specify CUDA_TOOLKIT_ROOT_DIR
Call Stack (most recent call first):
  /nix/store/xpx3wyc4ws1blif4y1zz92ys8gf2y3wz-opencv-4.9.0/lib/cmake/opencv4/OpenCVConfig.cmake:86 (find_package)
  /nix/store/xpx3wyc4ws1blif4y1zz92ys8gf2y3wz-opencv-4.9.0/lib/cmake/opencv4/OpenCVConfig.cmake:108 (find_host_package)
  CMakeLists.txt:62 (find_package)


-- Configuring incomplete, errors occurred!
```

### Additional context

It seems like the opencv package is still using the old `FindCUDA` module which is unable to find cuda somehow. 

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.13.1-zen1, NixOS, 25.05 (Warbler), 25.05.20250216.d74a233`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.12`
 - nixpkgs: `/nix/store/k8nkf470zpidpa5nh76lh2x6rxfzpwa4-source`

### Notify maintainers

@cole-h 

Not sure if I need to tag also cuda maintainers and/or opencv maintainers.

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/opencv/4.x.nix
+++ b/pkgs/development/libraries/opencv/4.x.nix
@@ -585,9 +585,13 @@ effectiveStdenv.mkDerivation {
       mkdir -p "$cxxdev/nix-support"
       echo "''${!outputDev}" >> "$cxxdev/nix-support/propagated-build-inputs"
     ''
+    # hard-wire CUDA_TOOLKIT_ROOT_DIR so FindCUDA sees the toolkit
     # remove the requirement that the exact same version of CUDA is used in packages
-    # consuming OpenCV's CMakes files
+    #   consuming OpenCV's CMakes files
     + optionalString enableCuda ''
+      sed -i '1s;^;set(CUDA_TOOLKIT_ROOT_DIR ${cudaPackages.cudatoolkit})\n;' \
+        "$out/lib/cmake/opencv4/OpenCVConfig.cmake"
+
       substituteInPlace "$out/lib/cmake/opencv4/OpenCVConfig.cmake" \
         --replace-fail \
           'find_host_package(CUDA ''${OpenCV_CUDA_VERSION} EXACT REQUIRED)' \

